### PR TITLE
Modernize some tests with Jest expect

### PR DESCRIPTION
**Pull request by karfau**
As proposed in https://github.com/xmldom/xmldom/issues/72#issuecomment-695050866 :

- use `expect(...)...` from jest instead of node assert
- when xmldom log output is visible during test, include it into the snapshot assertion by using `getTestParser`
- switch `var` to `const` inside tests

### CODE REVIEW

**Comment by brodybits** [comment]
Hey @karfau this looks really awesome so far, in general. I am just curious if this has any effect on the Stryker mutation testing score?

Also, please feel free to make multiple changes in the same PR if it helps unblock your work. I think it should be no problem for me or us to split things up later, if needed.

**Comment by karfau** [comment]
@brodybits Since it's already quite some stuff to review, I would prefer to get some feedback already.
Or we could even land it as is and I continue with the remaining tests afterwards.
(Commits are mostly split by transferred file currently, not sure that helps a lot though.)

What do you think?

File: test/parse/file.test.js
Comment (brodybits) [inline test/parse/file.test.js:4]: nit:

```suggestion

const { DOMParser } = require('../../lib/dom-parser')
```

File: test/parse/namespace.test.js
Comment (brodybits) [inline test/parse/namespace.test.js:3]: another nit:

```suggestion
const { DOMParser } = require('../../lib/dom-parser')
```

**Comment by brodybits** [comment]
@karfau welcome to the team!

I will review later today or tonight. Please bug me in case I forget.

I did take the liberty to update the title. Please feel free to change it again or change it back if you think it is needed.

I may have one more nit coming soon.

File: test/parse/node.test.js
Comment (brodybits) [inline test/parse/node.test.js:168]: last nit for now (hopefuly): I find the formatting of this statement to be a bit ugly. I would do something like this:

```suggestion
		const output = new DOMParser().parseFromString(source, 'text/xml').toString()
		expect(output).toBe(source)
```

Comment (karfau) [reply test/parse/node.test.js]: :+1: I will also check if there are other places where "arrange, act, assert" are not clearly split by (empty) lines.

**Comment by karfau** [comment]
@brodybits What is you preference: Landing and create a "part 2" PR or continue to add stuff to this one?

**Review by brodybits — COMMENTED** [review]
I left some questions, nits, and other comments. I skimmed through the last few test modules but the approach seemed OK to me.

I think I would favor that we take care of these comments now and consider making a "part 2" PR if we find any other things that need to be "modernized".

File: test/error/error.test.js
Comment (brodybits) [reply test/error/error.test.js]: This nit seems to be already resolved, my bad.

File: test/error/xml-error.test.js
Comment (karfau) [reply test/error/xml-error.test.js:7]: It's redundant.

File: test/html/normalize.test.js
Comment (karfau) [reply test/html/normalize.test.js:64]: They will just remain as is, but we can make this obvious. :+1:

File: test/error/error.test.js
Comment (karfau) [reply test/error/error.test.js:17]: Yes, but I think it's as simple as it can get.

We could of course introduce a small module that holds the most common mime types, or at least the ones we test for. Or even add some devDependency for that, but is it really worth it?

If you think it is I would prefer to do that change in a separate PR.

File: test/error/xml-error.test.js
Comment (karfau) [reply test/error/xml-error.test.js:5]: The naming of those tests suites is as weird as the grouping of the tests is (most of the times.
That's why I didn't pay much attention to all cases yet.

This file is a good example: it's called xml-errors nd it's mostly demonstrating that xml and html are treated differently.

`errorHandle[r]` is also used in the main describe of `error/error.test.html` which is really about the implementation of the `errorhandler` option that can be passed to `DOMParser`.

So maybe this describe should rather be `'xml vs html'`?

File: test/error/xml-error.test.js
Comment (karfau) [reply test/error/xml-error.test.js:19]: no, they differ in the mime type, I'll make that more obvious by using `it.each`.

File: test/error/error.test.js
Comment (karfau) [reply test/error/error.test.js:6]: Hm, the difference here is not the usage but the content.
- Parsing `XML_ERROR` will trigger a call to `errorHandler.error`
- Parsing `XML_ERROR_AND_WARNING` will trigger a call to `errorHandler.error` and `errorHandler.warning`

We curently only have one test where the error handler rethrows and it's proven that that error is not caught, but we could of cause add more tests where the document to test should only trigger `errorHandler.error`.

(I don't care about the ordering a lot, but with the current naming it's alphabetical.)

File: test/error/xml-error.test.js
Comment (karfau) [reply test/error/xml-error.test.js:45]: Good catch.
We already have the following tests in `test/html/normalize.test.js`:
- `<html test="a<b && a>b && \'&amp;&&\'"/>`
- `<div test="a<b&&a< c && a>d"></div>`
- `<div a=& bb c d=123&&456/>`
- `<div a=& a="&''" b/>`

So we could just add `&"/>` there if we think it adds any value, but I think it doesn't. So I would drop it. What do you think?

File: test/html/normalize.test.js
Comment (karfau) [reply test/html/normalize.test.js:40]: If you look at the snapshots that's what is happening. If at any point in the future an error is raised for any of those the test will fail. And since the `errors` dict is spread, there is no trace of it in the snapshot.

File: test/html/normalize.test.js
Comment (karfau) [reply test/html/normalize.test.js:63]: I don't hink it's required. it will jsut increase the test runtime and in cae of a test failure this single test will tell you which ones don't work.

File: test/parse/locator.test.js
Comment (karfau) [reply test/parse/locator.test.js:27]: I think tests should be as specific as possible, so they the lowest amount of tests fail when something goes wrong.

This one is a good example: It verifies that empty lines are also counted by the locator.

Would we also check the columnNumber (which is covered by other tests) it would just fail together with other tests that verify them, but there is no need for it to fail of the line number for empty lines is still correct.

File: test/parse/locator.test.js
Comment (karfau) [reply test/parse/locator.test.js:95]: As described in my comment on line 87 the tests doesn't yield any errors or warnings although the XML is not well formed. I made this explicit by also asserting on no errors being present.

File: test/parse/namespace.test.js
Comment (karfau) [reply test/parse/namespace.test.js:27]: I'm not sure why, do you want me to put an empty line between every assertion?
So far I only introduced empty lines between "arrange, act, assert", and I'm aware it's a bad practice t have multiple assertions per test.

So if you want I would rather split the test into multiple ones.

File: test/html/normalize.test.js
Comment (karfau) [reply test/html/normalize.test.js:62]: I was wrong, the `&` sign will of course be converted to `&amp;`. I added the test.

**Comment by karfau** [comment]
> I think I would favor that we take care of these comments now and consider making a "part 2" PR if we find any other things that need to be "modernized".

It's not about part 2 of modernizing. The "main goal" of this PR was to actually get rid of the custom assertions, and the whole directory `test/dom` (7 suites, 23 tests) has not been converted yet.
I'm fine with finishing the conversion in this PR or in a "part 2". Would love to know if you preferr we land this first.

File: test/error/error.test.js
Comment (brodybits) [reply test/error/error.test.js:6]: ok

File: test/error/error.test.js
Comment (brodybits) [reply test/error/error.test.js:17]: OK. I would love to avoid hard-coded strings, but this is probably not worth changing now.

**Comment by brodybits** [comment]
I took the liberty to change the title again. I would probably favor converting `test/dom` in a separate PR. But not so strongly, since it does not seem to be a really large amount to convert in `test/dom`. A side point is that we (I) did already start using Jest expect in recent PR #126.

Most of the dispositions make sense to me. I will try to give this another read-through today.

File: test/error/error.test.js
Comment (brodybits) [inline test/error/error.test.js:7]: I wonder if this is something that we should "restore" testing?

File: test/error/xml-error.test.js
Comment (brodybits) [reply test/error/xml-error.test.js:5]: Makes sense to me

File: test/error/xml-error.test.js
Comment (brodybits) [reply test/error/xml-error.test.js:7]: But I don't see where else p.parseFromString('', ...) is tested. Am I missing something?

File: test/error/xml-error.test.js
Comment (brodybits) [reply test/error/xml-error.test.js:37]: I guess it would not add much value.

I would probably keep this case, but leave it up to you.

File: test/html/normalize.test.js
Comment (brodybits) [reply test/html/normalize.test.js:61]: OK.

At some point I hope we can test all entries in `lib/entities.js`.

File: test/parse/big-file-performance.test.js
Comment (brodybits) [inline test/parse/big-file-performance.test.js:80]: Why remove this whole test?

File: test/parse/locator.test.js
Comment (brodybits) [reply test/parse/locator.test.js:27]: Got it, ok.

File: test/parse/locator.test.js
Comment (brodybits) [reply test/parse/locator.test.js:95]: Got it.

In this case, shouldn't we assert that there are no errors in some of the other tests as well?

File: test/error/xml-error.test.js
Comment (karfau) [reply test/error/xml-error.test.js:7]: https://github.com/xmldom/xmldom/blob/be1f8d5e7953078160af81672e8cf8344d47744c/test/parse/simple.test.js#L91

File: test/parse/locator.test.js
Comment (brodybits) [inline test/parse/locator.test.js:18]: It looks like you fixed a piece of the XML, which is good. But I think it would be a nice to test a piece of XML with `=` missing as well.

File: test/parse/locator.test.js
Comment (brodybits) [reply test/parse/locator.test.js:27]: And maybe add a comment here?

File: test/error/xml-error.test.js
Comment (karfau) [reply test/error/xml-error.test.js:37]: Then I'll drop it as "redundant".

File: test/parse/locator.test.js
Comment (karfau) [reply test/parse/locator.test.js:95]: I'm quite confident that I did it for all cases that I thought it makes sense for some reason, but will double check and maybe add a related comment or some README ...

File: test/parse/namespace.test.js
Comment (brodybits) [reply test/parse/namespace.test.js:27]: OK. I think this test is good for now. We can always improve it later.

File: test/parse/big-file-performance.test.js
Comment (karfau) [reply test/parse/big-file-performance.test.js:80]: Here are my reasons (should have highlighted them in advance): 
It was testing that the "performance" (measured in execution time) is better then two other libraries that are not listed in our devDependencies and that are also no longer actively maintained. It was also one of the slowest tests overall.

If I remember correctly, it was doing the comparison by modifying the prototype of DOMParser, so it modified global state.

File: test/parse/locator.test.js
Comment (karfau) [reply test/parse/locator.test.js:18]: :eyes: :+1: 
I think when I "fixed the XML" I discovered that the same XML was also present in one of the error test files. But I removed that part later since it was not used for testing anywhere. So this is related to your other comment about converting it into a test.

I will look into it again.

**Comment by karfau** [comment]
FYI: I will not be able to work on it again from beginning of next week.

File: test/parse/big-file-performance.test.js
Comment (brodybits) [reply test/parse/big-file-performance.test.js:80]: I would favor using `xit` or `xdescribe` instead, and maybe do this in a separate PR.

**Comment by brodybits** [comment]
Hi @karfau I did not understand your last comment. But thanks for all of your work on this, really good in general. Most of my comments have been minor nits, anyways. Anything we do not fix here can always be fixed in a future PR, if needed.

I leave it up to you to finish rework and merge of this PR, whenever you are ready. I am thinking to make the 0.4.0 release once this is merged.

**Comment by karfau** [comment]
Sorry my sentence was indeed confusing, updated it. I don't have access to my computer over the weekend, I'll continue to work on it when I'm back :)

File: test/error/xml-error.test.js
Comment (karfau) [reply test/error/xml-error.test.js]: Done, renamed it.

File: test/parse/locator.test.js
Comment (karfau) [reply test/parse/locator.test.js:95]: I think I will keep this work as part of checking for more duplicates and potentially rearranging them better as discussed in #72 

Also since I want to get this PR done.

File: test/error/error.test.js
Comment (karfau) [reply test/error/error.test.js:7]: Done

File: test/parse/locator.test.js
Comment (karfau) [reply test/parse/locator.test.js:18]: The one with missing `=` is now part of `xml-error.test.js`

File: test/parse/locator.test.js
Comment (karfau) [reply test/parse/locator.test.js:30]: Done

**Comment by karfau** [comment]
OK, I think everything I would like to do in this PR is done.

Feel free to check if it can land or add more comments if needed. Will check when I'm back.

**Review by brodybits — APPROVED** [review]
I think this is good to go. Please merge it whenever you can.

**Comment by brodybits** [comment]
Thanks to @karfau for the work on this. Did you mean to merge this with a merge commit?

**Pull request merged**
